text/prod-loader.md

```diff
@@ -0,0 +1,133 @@
+* Date: November 18th, 2016
+* Title: A minimal production loader
+
+# Summary
+
+Minimize the amount of Loader code that runs in production to only what is necessary to support module resolution, circular dependencies, and dynamic loading.
+
+# Motivation
+
+I did a [performance audit](https://github.com/stealjs/stealjs/issues/20) on Steal to see what shortcomings we currently have. The goal is to achieve 3 second load times on 3G with a mid-range phone.
+
+Based on this audit there are the falling problems with *steal.production.js*:
+
+* The total budget of JavaScript is 100k. steal.production.js accounts for 21k, so a little over 1/5th of the total JavaScript budget. This doesn't include the npm plugin which is also packed into the main bundle.
+* steal.production.js takes 176ms to execute on a mid-range phone.
+
+The goal of this change will be to:
+
+* Limit the amount of Steal code that needs to run in production to under 10k. I need to compare to other bundlers to see if this is achievable or not, but my guess is that it is.
+* Support circular dependencies.
+* Support progressive loading.
+* Support loading as async script tags so that multiple bundles can be loaded in parallel:
+
+```html
+<script async src="./path/to/prod.js"></script>
+<script async src="./path/to/bundle-a.js"></script>
+<script async src="./path/to/bundle-b.js"></script>
+```
+
+# Impact
+
+This will be small change to the affecting repositories and most work will be done in a new project. This would reflect a *minor* version change as it would introduce a new feature with no backwards compatibility concerns.  In the future this loader could become the default that steal-tools uses, which would require a *major* version change but there are no plans to do that until it has been used in the wild for a bit behind a flag.
+
+The following projects *will* be affected:
+
+* **transpile** - In the below design I lay out a new bundling format that this new loader will use. Transpile will need to be updated in order to transpile to this format.
+* **steal-tools** - Will need to be updated to accept a flag that will use this new loader.
+
+# Design
+
+> Note that API names here aren't important and can change, I'm using steal as the example.
+
+Since the goal of this RFC is a loader that is *as small as possible* it will only support the minimal amount of configuration needed to for progressive loading bundles.
+
+## Module ids replace module names
+
+**transpile** will transpile to a different format, let's call it **_steal_** for now, and will omit something like:
+
+```js
+steal([0, 7, 12, 34], function(exports, bar, baz, qux){
+
+});
+```
+
+Notice that instead of module names there are integers. This is to save space by omitting the long strings. During the build each module name will be given an integer id.
```

Comment:
Here's my first take of the results: https://gist.github.com/matthewp/980536eb793478e7bb0bb4f8e629c8ce

The readme explains the methodology. Results are likely skewed too strongly in favor of integer ids. The reason (I think), is that because the modules don't all have different function bodies, the difference between string module names and integer ids is the biggest difference between the bundles.

So I'm going to try and generate some random JavaScript function bodies so that each module has different code (thus creating more work for gzip) and see how that goes.